### PR TITLE
quic: add downstream config knob to enable/disable early data

### LINKED ISSUE #19943: Make enabling early-data configuration in HTTP/3
Add a config knob to set QuicSSLConfig::early_data_enabled accordingly.

### LINKED ISSUE #20109: quic: add a config knob for congestion control algorithm
Right now the default is BBR and enabled by hard coding a QUICHE reloadable flag in 
#19349 quic: enable BBR congestion control. But we want it to be configurable, especially distinguishable between upstream and downstream. 
As discussed with @ggreenway, the preferred default value should be CUBIC, in consideration of https://en.m.wikipedia.org/wiki/TCP_congestion_control#TCP_BBR.

### PR DESCRIPTION
**Pull request by danzh2010**
Commit Message: add envoy.extensions.transport_sockets.quic.v3.QuicDownstreamTransport.enable_early_data to turn on/off early data support in EnvoyQuicServerSession.

Risk Level: low, behavior change to h3 downstream handshake is behind the new config knob.
Testing: added new integration tests
Docs Changes: N/A
Release Notes: noted about the new config knob `transport_sockets.quic.v3.QuicDownstreamTransport.enable_early_data` 
Fixes #19943

### CODE REVIEW

File: api/envoy/extensions/transport_sockets/quic/v3/quic_transport.proto
Comment (htuch) [inline api/envoy/extensions/transport_sockets/quic/v3/quic_transport.proto:26]:
```suggestion
  // If false, QUIC will tell TLS to reject any early data and to stop issuing 0-RTT credentials with resumption session tickets. This will prevent clients from attempting 0-RTT requests.
```

File: api/envoy/extensions/transport_sockets/quic/v3/quic_transport.proto
Comment (htuch) [inline api/envoy/extensions/transport_sockets/quic/v3/quic_transport.proto:28]: I would be fine with this being `bool disable_early_data` as well, but up to you.

**Review by htuch — COMMENTED** [review]
API LGTM

**Comment by htuch** [comment]
@RyanTheOptimist for QUIC

**Comment by KfreeZ** [comment]
noticed another issue #19542 requiring the 0-rtt config, is earlydata knob in this PR also represent the 0-rtt config?

**Comment by danzh2010** [comment]
> noticed another issue #19542 requiring the 0-rtt config, is earlydata knob in this PR also represent the 0-rtt config?

That issue is talking about an upstream config knob. So not related to this one.

File: api/envoy/extensions/transport_sockets/quic/v3/quic_transport.proto
Comment (RyanTheOptimist) [reply api/envoy/extensions/transport_sockets/quic/v3/quic_transport.proto:28]: Sounds good to me. If earlier data is enabled by default, then having to explicitly set `disable_early_data = true` seems to express the right semantics.

File: source/common/quic/quic_transport_socket_factory.cc
Comment (RyanTheOptimist) [inline source/common/quic/quic_transport_socket_factory.cc:20]: I think this could also be: `!quic_transport.has_enable_early_data() || quic_transport.enable_early_data().value()`

File: source/common/quic/quic_transport_socket_factory.cc
Comment (RyanTheOptimist) [inline source/common/quic/quic_transport_socket_factory.cc:21]: Could we have a unit test of createTransportSocketFactory which verifies the expected value of the returned factory's enable_early_data in which we pass in quic_transport with no value, true and false for enable_early_data?

File: test/integration/quic_http_integration_test.cc
Comment (RyanTheOptimist) [inline test/integration/quic_http_integration_test.cc:1110]: If I'm reading this right, it looks like both tests end up with early data rejected. Is there any way to get a test where we accept early data?

**Review by RyanTheOptimist — COMMENTED** [review]
Nice!

File: api/envoy/extensions/transport_sockets/quic/v3/quic_transport.proto
Comment (htuch) [reply api/envoy/extensions/transport_sockets/quic/v3/quic_transport.proto:28]: Also, please add version history update for this knob. Thanks.

File: source/common/quic/quic_transport_socket_factory.cc
Comment (danzh2010) [reply source/common/quic/quic_transport_socket_factory.cc:21]: done

File: api/envoy/extensions/transport_sockets/quic/v3/quic_transport.proto
Comment (danzh2010) [reply api/envoy/extensions/transport_sockets/quic/v3/quic_transport.proto:28]: Using "enable" rather than "disable" here is to be consistent with QuicSSLConfig::early_data_enabled.

**Comment by danzh2010** [comment]
needs api stamp again @htuch

**Review by htuch — COMMENTED** [review]
/lgtm api

**Comment by danzh2010** [comment]
Ping @RyanTheOptimist

**Review by RyanTheOptimist — APPROVED** [review]

**Comment by RyanTheOptimist** [comment]
While I'm able to merge HTTP/3 changes without cross-organization review, since this also includes an API change, I'd like to tick that box. I think @ggreenway has had thoughts in this space.

**Comment by ggreenway** [comment]
@danzh2010 I think you linked the wrong issue to fix. Did you mean #19943?

File: source/common/quic/envoy_quic_server_session.cc
Comment (ggreenway) [inline source/common/quic/envoy_quic_server_session.cc:186]: The other place in the code-base I see this cast it is a dynamic_cast, in `EnvoyQuicProofSource::getTlsCertConfigAndFilterChain`. Should this be the same?

File: source/common/quic/quic_transport_socket_factory.cc
Comment (ggreenway) [inline source/common/quic/quic_transport_socket_factory.cc:20]: `PROTOBUF_GET_WRAPPED_OR_DEFAULT` would read easier.

**Review by ggreenway — CHANGES_REQUESTED** [review]
Looks good overall.

/wait

File: source/common/quic/quic_transport_socket_factory.cc
Comment (danzh2010) [reply source/common/quic/quic_transport_socket_factory.cc:20]: done

**Review by ggreenway — DISMISSED** [review]

**Comment by ggreenway** [comment]
CI is failing, and there's a merge conflict.

/wait

**Comment by danzh2010** [comment]
> CI is failing, and there's a merge conflict.
> 
> /wait

Sync'ed with main now.

**Comment by danzh2010** [comment]
Ping for merge

**Comment by RyanTheOptimist** [comment]
> Ping for merge

I'll be happy to merge once @ggreenway approves.

**Comment by RyanTheOptimist** [comment]
Oh, @ggreenway previously approved. Sorry, I'll merge now.

**Review by RyanTheOptimist — APPROVED** [review]

**Pull request merged**
